yacl/crypto/primitives/psu/krtw19_psu.cc

```diff
@@ -0,0 +1,228 @@
+
+constexpr size_t NUM_BASE_OT{128ul};
+constexpr size_t NUM_INKP_OT{512ul};
+
+static std::random_device rd;
+static std::mt19937 gen(rd());
+
+struct U128Hasher {
+  size_t operator()(const uint128_t& x) const {
+    auto hash = Blake3_128({&x, sizeof x});
+    size_t ret;
+    std::memcpy(&ret, &hash, 1);
+    return ret;
+  }
+};
+
+auto HashInputs(const std::vector<uint128_t>& elem_hashes, size_t count) {
+  size_t num_bins = std::ceil(count * ZETA);
+  std::vector<std::vector<uint128_t>> hashing(num_bins);
+  for (auto elem : elem_hashes) {
+    auto hash = U128Hasher{}(elem);
+    hashing[hash % num_bins].emplace_back(elem);
+  }
+  return hashing;
+}
+
+uint64_t Evaluate(const std::vector<uint64_t>& coeffs, uint64_t x) {
+  uint64_t y{coeffs.back()};
+  for (auto it = std::next(coeffs.rbegin()); it != coeffs.rend(); ++it) {
+    y = GfMul64(y, x) ^ *it;
+  }
+  return y;
+}
+
+auto Interpolate(const std::vector<uint64_t>& xs,
+                 const std::vector<uint64_t>& ys) {
+  YACL_ENFORCE_EQ(xs.size(), ys.size(), "Sizes mismatch.");
+  size_t size{xs.size()};
+  std::vector<uint64_t> L_coeffs(size);
+  for (size_t i{}; i != size; ++i) {
+    std::vector<uint64_t> Li_coeffs(size);
+    Li_coeffs[0] = ys[i];
+    uint64_t prod{1};
+    for (size_t j{}; j != size; ++j) {
+      if (xs[i] != xs[j]) {
+        prod = GfMul64(prod, xs[i] ^ xs[j]);
+        uint64_t sum{};
+        for (size_t k{}; k != size; ++k) {
+          sum = std::exchange(Li_coeffs[k], GfMul64(Li_coeffs[k], xs[j]) ^ sum);
+        }
+      }
+    }
+    for (size_t k{}; k != size; ++k) {
+      L_coeffs[k] ^= GfMul64(Li_coeffs[k], Inv64(prod));
+    }
+  }
+  return L_coeffs;
+}
+
+}  // namespace
+
+void KrtwPsuSend(const std::shared_ptr<yacl::link::Context>& ctx,
+                 const std::vector<uint128_t>& elem_hashes) {
+  ctx->SendAsync(ctx->NextRank(), SerializeUint128(elem_hashes.size()),
+                 "Send set size");
+  size_t peer_count =
+      DeserializeUint128(ctx->Recv(ctx->PrevRank(), "Receive set size"));
+  auto count = std::max(elem_hashes.size(), peer_count);
+  if (count == 0) {
+    return;
+  }
+  // Step 1. Hashes inputs
+  auto hashing = HashInputs(elem_hashes, count);
+
+  // Step 2. Prepares OPRF
+  KkrtOtExtReceiver receiver;
+  size_t num_ot{hashing.size() * BIN_SIZE};
+  auto choice = RandBits(NUM_BASE_OT);
+  auto base_ot = BaseOtRecv(ctx, choice, NUM_BASE_OT);
+  auto store = IknpOtExtSend(ctx, base_ot, NUM_INKP_OT);
+  receiver.Init(ctx, store, num_ot);
+  receiver.SetBatchSize(BATCH_SIZE);
+
+  std::vector<uint128_t> elems;
+  elems.reserve(num_ot);
+  size_t oprf_idx{};
+  for (size_t bin_idx{}; bin_idx != hashing.size(); ++bin_idx) {
+    if (bin_idx % NUM_BINS_PER_BATCH == 0) {
+      receiver.SendCorrection(
+          ctx, std::min(BATCH_SIZE, (hashing.size() - bin_idx) * BIN_SIZE));
+    }
+    hashing[bin_idx].resize(BIN_SIZE);
+    std::shuffle(hashing[bin_idx].begin(), hashing[bin_idx].end(), gen);
```

Comment:
Since `hashing[bin_idx]` contains only hashed values, you may use `std::sort` (faster) instead of `std::shuffle`. Many paper states that hash and sort equals to shuffle assuming hash function is an instantiation of Random Oracle.